### ISSUE TITLE
Start considering configure serials for xdg-shell

### ISSUE THREAD

**Issue by ammen99** [WayfireWM/wayfire#995]
Currently, we ignore the xdg-shell configure serials which can cause quite a lot of problems. WIth #973 there is some minimal support but the whole maximize/fullscreen/resize implementation needs to be updated to achieve "perfect frames". An example of how things should work with maximizing:

- client or plugin requests resize
- grid (or as a fallback core) gets this request, calculates geometry and sends a configure (with serial N) to the client, with maximized state and desired width/height
- time passes
- client commits a new buffer after ACK-ing configure N.
- Wayfire/Grid/whatever starts the maximize animation from previous to new buffer (for ex. crossfade animation, or wobbly)

How things work now:

- client or plugin requests resize
- grid (or as a fallback core) gets this request, calculates geometry and sends a configure (with serial N) to the client, with maximized state and desired width/height. Immediately start maximize animation
- time passes, animation runs, wrong visual result since x/y of the view has been modified already.
- client commits a new buffer after ACK-ing configure N.

Note that implementing cross-fade in Grid instead of the current simple move+resize animation will still give a much better visual result when the client resizes almost immediately (maybe within 1-3 frames).

On the API side, we'll have to add quite a few changes. First of all, we'll need to expose a new set of functions which just send the configure to the request, and a new set of signals for when the client ACKs them (or maybe just expose pre-commit and commit events for views). Then, the relevant plugins will need to be updated to use that. Given it will be a breaking  change, we should do this before 1.0.

For xwayland, we do not have serials. This means we will have to try and guess what the client has done. For example, maching fullscreen/maximized state, window size (changes), etc.

**Comment by soreau** [comment]
Since synchronizing the resize with positioning of a surface could be a useful thing in different cases, it might make sense to generalize this code so the delicate parts are internal and not duplicated.

**Comment by ammen99** [comment]
Moving this to the next milestone, because an implementation would contain a lot of breaking changes. I have been thinking about using a transaction model like Sway and River do. A bit of brain dump below, of course questions/suggestions are welcome, I have not started working on the code yet.

We should make it possible to change the following view properties in an atomic fashion:
  - Position
  - Size
  - Fullscreen/Tiled/Minimized
  - Maybe activated or not
  - Anything else??

Here atomic means for ex: A plugin wants to `set_geometry()` on the view. This involves changing both size and position. Currently, we change the position, at when the client commits, we change the size too. If we go atomic, Wayfire would not apply the change in position until the size changes too. This would allow smoother transitions in some cases I believe (for ex. using Grid to put a slow-resizing view in a slot).

I have also been thinking, in the context of tiling plugins and perhaps other use-cases (could eventually lead to a possibility of decorations rendered by a client) we want to apply changes to properties to multiple views atomically (i.e apply changes once all of the views are ready). How should the interface for plugins look like? Maybe this:

- There is an obvious need for splitting the aforementioned states out of the view class. Would need a new struct `wf::view_state_t`. A view should also have `wf::view_interface_t::current_state, pending_state` to differentiate between currently displayed state and state which has been requested, but not applied yet. This could be useful information to plugins, the inability to know the difference has caused quite a lot of bugs so far (which we have worked around).

- We need a `transaction_t` type which holds new states for one or more views. To change the state of a view or a group of views, a plugin would need to build a `transaction_t`, submit it to the output (`wf::output_t::submit_transaction` or smth like that), get a serial id for that transaction, and potentially wait until it either 1) times out, if the client is unresponsive or 2) is accepted by the client(s)

- We will need `uint32_t wf::view_interface_t::request_state(const view_state_t& next), apply_pending_state()` to send a new pending state to the client and to set the current state. This is however not an API for plugins: everything will need to go to the output transaction, and it manages the requests and applies the results, when all views in a transaction have updated successfully.

- The API would be a bit cumbersome to use in simpler cases (which is actually most of the plugins). We could provide some helper functions to make it easier, perhaps with the builder pattern:
```cpp
view->new_state()
    .set_size(...)
    .set_fullscreen(true)
    .submit();
```
Plugins would then have a simple way to request a few changes to a single view, without being involved with `output_t` and `transaction_t`.

**Comment by ammen99** [comment]
Some more thoughts on the transaction system, inspired by discussions in #sway-devel:

- Transactions can have committed and pending state. Pending means still accumulating new state, committed means already sent to client. We can have multiple committed transactions, as long as the participating views are disjoint. If we get a new request/transaction for a view that is already in a committed transaction, we should create a new pending one and accumulate any such requests there, until the committed transaction(s) are done, at which point the pending transaction is committed.

- Rendering state such as output geometry, surface buffers, view transforms should not be duplicated in a current/pending combination. Instead, they will continue to reflect the **current** view state, without committed or pending transactions applied to the view.

- There is an upcoming wlroots infrastructure which will make it easy to lock surface state and keep track of when a view ACKs the new state https://github.com/swaywm/wlroots/issues/2957. We should make good use of this functionality.

- Contrary to what I originally planned, it makes the most sense to track views on a global scale, not per-output. This way, we could easily have frame-perfect operations involving moving a view between outputs. We may then also need to make the view's output part of the pending or current state.

- There are various signals like `view-geometry-changed`, etc. They will no longer make that much sense, we should let plugins know when transactions are ready, as multiple views may be changing their position or other state at the same time.

- In the above line of thinking, changing the active workspace should involve a transaction moving all views together.

- On the implementation side, probably the easiest way is something like Sway's implementation: each transaction contains several instructions, each of which is a basic view operation (e.g move to an output, set geometry, set fullscreen, etc.). An instruction interface could look like this:
```cpp
class set_geometry_instruction_t : public transaction_instruction_t
{
<constructor>
// Request the given state from the client
void commit();
// Apply changes to the compositor state, these changes should NOT emit any signals or trigger plugins except compositor view implementations
void apply();
// Check a view commit for completeness
// We need a commit event structure which is enough for what both client and compositor views need
bool check(<commit event>);
};
```

To support all of this, `view_interface_t` will also have to be augmented:

```cpp
struct view_state_t {
// all kinds of state here, including current output, wm geometry, etc.
};

// This belongs to view-impl.hpp
class view_impl {
// Implementation of all the functions for changing state, moving to another output, etc. 
};

class view_interface_t .. {
// Allow only transactions to change the view state non-atomically
view_impl& get_impl(); // Can be accessed only by core

// Get current state (as will currently be presented to the user)
const view_state_t& current();
// Get scheduled state (both committed and pending transactions, so that plugins know where the view will end up)
const view_state_t& pending();

// Same as before:

// Returns geometry from current
 get_output_geometry()
// Returns geometry from current, for compatibility, but deprecated
 get_wm_geometry()
// Returns current state
get_wlr_buffer()
// Returns current output, for compatibility, but deprecated
 get_output()
// ...
};
```

**Comment by Javyre** [comment]
Would all the old view `get_*` methods be deprecated or just the two with the `// ... but deprecated` comment in the above example?

**Comment by ammen99** [comment]
> Would all the old view `get_*` methods be deprecated or just the two with the `// ... but deprecated` comment in the above example?

My PR is rather functional and feature-complete (wrt api changes), so you should check it out :) The recommended way is to use `state()` or `pending()` for almost everything, the other functions are the for backwards compatibility and also because they are shorter ... I'm not yet sure whether they should be eventually removed or not.

**Comment by ammen99** [comment]
Finally fixed in #1704
